### PR TITLE
tool: add detector for multiple new reinitializers

### PR DESCRIPTION
**Pull request by QiuhaoLi**
When upgrading a proxy's implementation with [reinitializer](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/utils/Initializable.sol#L152), developers could errorly write multiple new reinitializers. If the new reinitializer with a higher version is called first, the reinitializer with a lower version can't be called later (you can't call a reinitializer with a lower version) and may leave uninitalized state variables.

This could lead to severe results in the real world. For example, the Ronin Network prepared two upgrades for their gateway contract and wrote two reinitializers (`initializeV3()` and `initializeV4()`). However, the developer calls the `initializeV4()` first and leaves the state variables `_operatorWeight` and `_totalOperatorWeight` uninitialized, which should be set in the `initializeV3()`. So attackers can bypass the vote weight and withdraw assets directly [[1]](https://rekt.news/roninnetwork-rektII/) [[2]](https://app.blocksec.com/explorer/tx/eth/0x2619570088683e6cc3a38d93c3d98899e5783864e15525d5f5810c11189ba6cb?line=2) [[3]](https://vscode.blockscan.com/ethereum/0xfc274ec92bbb1a1472884558d1b5caac6f8220ee).

ronin-exploit1

The best practice for this issue is never to introduce more than one reinitializer in one updated implementation. This PR will add a detector to the `slither-check-upgradeability` to check if multiple new reinitializes are introduced and report a low-severity warning. For example:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.24;
import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
contract Counter is Initializable {
    uint256 public x;

    function initialize(uint256 _x) public initializer {
        x = _x;
    }

    function initializeV2(uint256 _x) public reinitializer(2) {
        x = _x;
    }

    function changeX() public {
        x++;
    }
}

contract CounterV2 is Initializable {
    uint256 public x;
    uint256 public y;
    uint256 public z;

    function initialize(uint256 _x) public initializer {
        x = _x;
    }

    function initializeV2(uint256 _x) public reinitializer(2) {
        x = _x;
    }

    function initializeV3(uint256 _y) public reinitializer(3) {
        y = _y;
    }

    function initializeV4(uint256 _z) public reinitializer(4) {
        z = _z;
    }

    function changeX() public {
        x = x + y + z;
    }
}
```

```bash
qiuhao@laptop:~/tmp/foundry_test/upgrade$ slither-check-upgradeability src/Counter.sol Counter --new-contract-name CounterV2
...
INFO:Slither:
CounterV2.initializeV3(uint256) (src/Counter.sol#33-35) multiple new reinitializers.
CounterV2.initializeV4(uint256) (src/Counter.sol#37-39) multiple new reinitializers.
Reference: https://github.com/crytic/slither/wiki/Upgradeability-Checks#multiple-new-reinitializers
qiuhao@laptop:~/tmp/foundry_test/upgrade$ 
```

### CODE REVIEW

**Comment by QiuhaoLi** [comment]
@0xalpharush Hi, could you help review this PR?

File: slither/tools/upgradeability/checks/initialization.py [hunk 1]
```diff
@@ -396,3 +400,96 @@ def _check(self):
...
+    }
+}
+```
+`CounterV2` has two reinitializers with new versions `2` and `3`. If `initializeV3()` is called first, the `initializeV2()` can't be called to initialize `y`, which may brings security risks.
+"""
+    # endregion wiki_exploit_scenario
+
+    # region wiki_recommendation
+    WIKI_RECOMMENDATION = """
+Do not use multiple reinitializers with higher versions in the updated contract.
```
Comment (0xalpharush) [inline slither/tools/upgradeability/checks/initialization.py:457]: Can we add what should be done instead? I think they should be combined into a single re-initializer per upgrade based off the description of the issue, right?

File: slither/tools/upgradeability/checks/initialization.py [hunk 2]
```diff
@@ -396,3 +400,96 @@ def _check(self):
...
+        initializer_funcs_v2 = _get_initialize_functions(contract_v2)
+        new_reinitializer_funcs = []
+        for fv2 in initializer_funcs_v2:
+            if not any((fv2.full_name == fv1.full_name) for fv1 in initializer_funcs_v1):
+                new_reinitializer_funcs.append(fv2)
+
+        results = []
+        if len(new_reinitializer_funcs) > 1:
+            for f in new_reinitializer_funcs:
+                info = [f, " multiple new reinitializers.\n"]
```
Comment (0xalpharush) [inline slither/tools/upgradeability/checks/initialization.py:492]:
```suggestion
                info = [f, " multiple new reinitializers which should be combined into one per upgrade.\n"]
```
Same general idea as https://github.com/crytic/slither/pull/2536/files#r1722628467

**Comment by 0xalpharush** [comment]
Can you run `make reformat && make lint` and fix the lint warnings please?

File: slither/tools/upgradeability/checks/initialization.py (hunk 1 above)
Comment (QiuhaoLi) [reply slither/tools/upgradeability/checks/initialization.py]: Yes, I added a suggestion, thanks.

**Comment by 0xalpharush** [comment]
I am trying to fix the CI failure in https://github.com/crytic/slither/pull/2537

**Comment by 0xalpharush** [comment]
Thanks @QiuhaoLi, these are great additions!

**Comment by QiuhaoLi** [comment]
@0xalpharush Thanks for the review :)

**Pull request merged**